File: chapter-13/ios-specific/EmbeddingDemo/ios/EmbeddingDemo/UserNameManager.swift
import Foundation

/**
 * Recipe 3: React Native → Native iOS Communication
 *
 * This native module receives method calls FROM React Native.
 * When the RN component calls UserNameManager.setUserName(text),
 * it arrives here. In a real app, you could update native UI,
 * post a notification, save to UserDefaults, etc.
 */
@objc(UserNameManager)
class UserNameManager: NSObject {

  @objc
  func setUserName(_ userName: String) {
    print("Received userName from RN: \(userName)")
    // In a real app: post NSNotification, update native UI, etc.
    NotificationCenter.default.post(
      name: NSNotification.Name("UserNameUpdated"),
      object: nil,
      userInfo: ["userName": userName]
    )
  }

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }
}

File: chapter-13/android-specific/EmbeddingDemo/android/app/src/main/java/com/embeddingdemo/UserNameManager.kt
package com.embeddingdemo

import android.util.Log
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod

/**
 * Recipe 3: React Native → Native Communication
 *
 * This native module receives method calls FROM React Native.
 * When the RN component calls UserNameManager.setUserName(text),
 * it arrives here. In a real app, you could update native UI,
 * save to SharedPreferences, trigger a broadcast, etc.
 */
class UserNameManager(reactContext: ReactApplicationContext) :
    ReactContextBaseJavaModule(reactContext) {

    override fun getName(): String = "UserNameManager"

    @ReactMethod
    fun setUserName(userName: String) {
        Log.d("UserNameManager", "Received userName from RN: $userName")
        // In a real app: update native UI, broadcast to other components, etc.
    }
}
